naming/dns_resolver_test.go

```diff
@@ -0,0 +1,304 @@
+func TestResolveFunc(t *testing.T) {
+	r, err := NewDNSResolver()
+	if err != nil {
+		t.Errorf("%v", err)
+	}
+	for _, v := range validAddrTc {
+		_, err := r.Resolve(v.addr)
+		if !reflect.DeepEqual(err, v.want) {
+			t.Errorf("Resolve(%s) = %v, want %v", v.addr, err, v.want)
+		}
+	}
+}
+
+//TODO(yuxuanli): Do we need to test with net.LookupHost, net.LookupSRV and real target? If not, delete this.
+var realTargetTc = []struct {
+	target string
+	want   []*Update
+}{}
+
+var fakeTargetTc = []struct {
+	target string
+	want   []*Update
+}{
+	{
+		"foo.bar.com",
+		[]*Update{newUpdate(Add, "1.2.3.4"+colonDefaultPort), newUpdate(Add, "5.6.7.8"+colonDefaultPort)},
+	},
+	{
+		"foo.bar.com:1234",
+		[]*Update{newUpdate(Add, "1.2.3.4:1234"), newUpdate(Add, "5.6.7.8:1234")},
+	},
+	{
+		"srv.ipv4.single.fake",
+		[]*Update{newUpdateWithMD(Add, "1.2.3.4:1234", "ipv4.single.fake")},
+	},
+	{
+		"srv.ipv4.multi.fake",
+		[]*Update{
+			newUpdateWithMD(Add, "1.2.3.4:1234", "ipv4.multi.fake"),
+			newUpdateWithMD(Add, "5.6.7.8:1234", "ipv4.multi.fake"),
+			newUpdateWithMD(Add, "9.10.11.12:1234", "ipv4.multi.fake")},
+	},
+	{
+		"srv.ipv6.single.fake",
+		[]*Update{newUpdateWithMD(Add, "[2607:f8b0:400a:801::1001]:1234", "ipv6.single.fake")},
+	},
+	{
+		"srv.ipv6.multi.fake",
+		[]*Update{
+			newUpdateWithMD(Add, "[2607:f8b0:400a:801::1001]:1234", "ipv6.multi.fake"),
+			newUpdateWithMD(Add, "[2607:f8b0:400a:801::1002]:1234", "ipv6.multi.fake"),
+			newUpdateWithMD(Add, "[2607:f8b0:400a:801::1003]:1234", "ipv6.multi.fake"),
+		},
+	},
+}
+
+var (
+	targetTc = realTargetTc
+)
+
+var hostLookupTbl = map[string][]string{
+	"foo.bar.com":      {"1.2.3.4", "5.6.7.8"},
+	"ipv4.single.fake": {"1.2.3.4"},
+	"ipv4.multi.fake":  {"1.2.3.4", "5.6.7.8", "9.10.11.12"},
+	"ipv6.single.fake": {"2607:f8b0:400a:801::1001"},
+	"ipv6.multi.fake":  {"2607:f8b0:400a:801::1001", "2607:f8b0:400a:801::1002", "2607:f8b0:400a:801::1003"},
+}
+
+var srvLookupTbl = map[string][]*net.SRV{
+	"_grpclb._tcp.srv.ipv4.single.fake": {newSRVRR("ipv4.single.fake", 1234)},
+	"_grpclb._tcp.srv.ipv4.multi.fake":  {newSRVRR("ipv4.multi.fake", 1234)},
+	"_grpclb._tcp.srv.ipv6.single.fake": {newSRVRR("ipv6.single.fake", 1234)},
+	"_grpclb._tcp.srv.ipv6.multi.fake":  {newSRVRR("ipv6.multi.fake", 1234)},
+}
+
+func updatesToSlice(updates []*Update) []Update {
+	res := make([]Update, len(updates))
+	for i, u := range updates {
+		res[i] = *u
+	}
+	return res
+}
+
+func testResolver(t *testing.T, freq time.Duration, slp time.Duration) {
+	for _, a := range targetTc {
+		r, err := NewDNSResolverWithFreq(freq)
+		if err != nil {
+			t.Fatalf("%v\n", err)
+		}
+		w, err := r.Resolve(a.target)
+		if err != nil {
+			t.Fatalf("%v\n", err)
+		}
+		var updates []*Update
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				u, err := w.Next()
+				if err != nil {
+					return
+				}
+				updates = u
+			}
+		}()
+		// Sleep for sometime to let watcher do more than one lookup
```

Comment:
Wouldn't a second call to Next() return an empty updates list, meaning this test would fail?

Why not call Next() in the main goroutine to make synchronization simper?